logwatch/logwatch.go

```diff
@@ -25,30 +27,101 @@ type LogNotification struct {
 // LogWatch is a test helper struct to monitor docker container logs for some patterns
 // and push their logs into Loki for further analysis
 type LogWatch struct {
-	t          *testing.T
-	log        zerolog.Logger
-	loki       *wasp.LokiClient
-	patterns   map[string][]*regexp.Regexp
-	notifyTest chan *LogNotification
-	containers []testcontainers.Container
-	consumers  map[string]*ContainerLogConsumer
-}
-
-// NewLogWatch creates a new LogWatch instance, with a Loki client
-func NewLogWatch(t *testing.T, patterns map[string][]*regexp.Regexp) (*LogWatch, error) {
-	loki, err := wasp.NewLokiClient(wasp.NewEnvLokiConfig())
-	if err != nil {
+	testName          string
+	log               zerolog.Logger
+	loki              *wasp.LokiClient
+	patterns          map[string][]*regexp.Regexp
+	notifyTest        chan *LogNotification
+	containers        []testcontainers.Container
+	consumers         map[string]*ContainerLogConsumer
+	logTargetHandlers map[LogTarget]HandleLogTarget
+}
+
+type LogContent struct {
+	TestName      string
+	ContainerName string
+	Content       []byte
+}
+
+type Option func(*LogWatch)
+
+// NewLogWatch creates a new LogWatch instance, with Loki client only if Loki log target is enabled (lazy init)
+func NewLogWatch(t *testing.T, patterns map[string][]*regexp.Regexp, options ...Option) (*LogWatch, error) {
+	l := logging.GetLogger(nil, "LOGWATCH_LOG_LEVEL").With().Str("Component", "LogWatch").Logger()
+	var testName string
+	if t == nil {
+		testName = NO_TEST
+	} else {
+		testName = t.Name()
+	}
+
+	logWatch := &LogWatch{
+		testName:          testName,
+		log:               l,
+		patterns:          patterns,
+		notifyTest:        make(chan *LogNotification, 10000),
+		consumers:         make(map[string]*ContainerLogConsumer, 0),
+		logTargetHandlers: getDefaultLogHandlers(),
+	}
+
+	for _, option := range options {
+		option(logWatch)
+	}
+
+	if err := logWatch.validateLogTargets(); err != nil {
 		return nil, err
 	}
-	l := logging.GetLogger(t, "LOGWATCH_LOG_LEVEL").With().Str("Component", "LogWatch").Logger()
-	return &LogWatch{
-		t:          t,
-		log:        l,
-		loki:       loki,
-		patterns:   patterns,
-		notifyTest: make(chan *LogNotification, 10000),
-		consumers:  make(map[string]*ContainerLogConsumer, 0),
-	}, nil
+
+	return logWatch, nil
+}
+
+func (m *LogWatch) validateLogTargets() error {
```

Comment:
@Tofel maybe it's better to make it a function (not method) that takes log targets? We could call it before creating `LogWatch` object in 58th line

Reply:
I don't see why not 👍

Reply:
that's ok. We can leave it then!